extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/configproperties/InterfaceConfigPropertiesUtil.java

```diff
@@ -0,0 +1,255 @@
+                        String defaultValueStr = nameAndDefaultValue.getDefaultValue();
+                        if (DotNames.OPTIONAL.equals(returnType.name())) {
+                            if (defaultValueStr != null) {
+                                /*
+                                 * it doesn't make to use @ConfigProperty(defaultValue="whatever") on a method that returns
+                                 * Optional
+                                 * since the result in this case isn't "optional", but there is always a value
+                                 */
+                                throw new IllegalArgumentException(
+                                        "Annotating a method returning Optional with @ConfigProperty and setting defaultValue is not supported. Offending method is "
+                                                + method.name() + " of interface" + ifaceDotName);
+                            }
+
+                            // use config.getOptionalValue to obtain the result
+
+                            Type genericType = determineSingleGenericType(returnType,
+                                    method.declaringClass().name());
+                            ResultHandle result = methodCreator.invokeInterfaceMethod(
+                                    MethodDescriptor.ofMethod(Config.class, "getOptionalValue", Optional.class, String.class,
+                                            Class.class),
+                                    config, methodCreator.load(fullConfigName),
+                                    methodCreator.loadClass(genericType.name().toString()));
+                            methodCreator.returnValue(result);
+                        } else {
+                            if (defaultValueStr != null) {
+                                /*
+                                 * When the default value is set, we need to use it if the property does not exist.
+                                 * So in this case we use config.getOptionalValue and check isPresent
+                                 */
+
+                                ReadOptionalResponse readOptionalResponse = createReadOptionalValueAndConvertIfNeeded(
+                                        generatedAbstractConfigPropertiesUtilClassName,
+                                        fullConfigName, returnType,
+                                        ifaceDotName, methodCreator, config);
+
+                                // just return the data if the Optional has data
+                                readOptionalResponse.getIsPresentTrue().returnValue(readOptionalResponse.getValue());
+
+                                // if the Optional does not contain data, we need to use the default value making sure we convert it first
+                                BytecodeCreator isPresentFalse = readOptionalResponse.getIsPresentFalse();
```

Comment:
Once we sort out some spec issues with MP-config (hopefully tomorrow) then the way to get the default value will be different from this: essentially what you're doing here will not work correctly because if a user forces a property to "empty" then you'll get the default value instead of the empty value.

A more correct approach that works today is to produce a config source which contains your default values at a low priority.  There's also a build item for this purpose: `io.quarkus.deployment.builditem.RunTimeConfigurationDefaultBuildItem`.

Reply:
That's great to know! I'll give it a shot

Reply:
Fixed